### PR TITLE
Add FSST as compression codec

### LINKED ISSUE #34246: Apply FSST compression for strings
See https://github.com/cwida/fsst

>FSST is a compression scheme focused on string/text data: it can compress strings from distributions with many different values (i.e. where dictionary compression will not work well). It allows random-access to compressed data: it is not block-based, so individual strings can be decompressed without touching the surrounding data in a compressed block. When compared to e.g. LZ4 (which is block-based), FSST further achieves similar decompression speed and compression speed, and better compression ratio.

### PR DESCRIPTION
**Pull request by Pelanglene**
Closes #34246

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Added FSST (Fast Static Symbol Table) as a new compression codec for string columns. 

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

---
### Modify your CI run:
**NOTE:** If your merge the PR with modified CI you **MUST KNOW** what you are doing
**NOTE:** Checked options will be applied if set before CI RunConfig/PrepareRunConfig step

#### Include tests (required builds will be added automatically):
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Unit tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with Analyzer
- [ ]  Add your option here

#### Exclude tests:
- [ ]  Fast test
- [ ]  Integration Tests
- [ ]  Stateless tests
- [ ]  Stateful tests
- [ ]  Performance tests
- [ ]  All with ASAN
- [ ]  All with TSAN
- [ ]  All with MSAN
- [ ]  All with UBSAN
- [ ]  All with Coverage
- [ ]  All with Aarch64
- [ ]  Add your option here

#### Extra options:
- [ ]  do not test (only style check)
- [ ]  disable merge-commit (no merge from master before tests)
- [ ]  disable CI cache (job reuse)

#### Only specified batches in multi-batch jobs:
- [ ]  1
- [ ]  2
- [ ]  3
- [ ]  4

### CODE REVIEW

**Comment by vdimir** [comment]
Ref https://github.com/ClickHouse/ClickHouse/issues/34246

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit 6526370f2e9e878c8ade881cc7c7b98b0b7df7c6 with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/62670/6526370f2e9e878c8ade881cc7c7b98b0b7df7c6/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>There's no description for the check yet, please add it to tests/ci/ci_config.py:CHECK_DESCRIPTIONS</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by alexey-milovidov** [comment]
Option 1: Expand the ICompressionCodec interface, add an optional Filter parameter to the decompression method—a structure containing, for example, required_substrings. The codec pretends to decompress all data but has the right to decompress unnecessary strings as empty. The filter is created in the query interpreter (more precisely, during query analysis) and then is carried through all data reading interfaces down to decompression.

Option 2: If a String (possibly Nullable, Array of String) uses the FSST codec, during deserialization, return a ColumnFSST instead of ColumnString. This is a new type of column that contains uncompressed bytes from one or more granules for subsequent lazy decompression. Only a few of its methods are implemented (filter, cut), whereas in almost all other cases, it must first be materialized into a full-fledged column.

Without these ways to accelerate, the codec does not make much sense - for raw performance/compression rate, it is strictly worse than LZ4 or ZSTD.

File: src/Compression/CompressionCodecFSST.cpp [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+            lens.push_back(cur_len);
+            rows.push_back(data);
+            data += cur_len;
+        }
+    }
+};
+
+void registerCodecFSST(CompressionCodecFactory & factory)
+{
+    auto codec_builder = [&](const ASTPtr & arguments) -> CompressionCodecPtr
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:134]: I guess we should restrict FSST to `String` and `FixedString` columns here. Check the `registerCodec` method src/Compression/CompressionCodecGorilla.cpp (which is a floating-point-value-only codec) how to do that.

Also, we should check that no further parameters were passed into `CODEC FSST()` when defined in SQL. Please see the `registerCodec` method in src/Compression/CompressionCodecGCD.cpp.

Both cases also need to have a negative SQL test, i.e. a test case in 02973_fsst_code_test_data.sql that expect an exception:

```sql
CREATE TABLE table_fsst_codec (n String CODEC(FSST('an_additional_argument'))) ENGINE = Memory; -- { serverError ILLEGAL_SYNTAX_FOR_CODEC_TYPE }
```

File: src/Compression/CompressionCodecFSST.cpp [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+class CompressionCodecFSST : public ICompressionCodec
+{
+public:
+    explicit CompressionCodecFSST() { setCodecDescription("FSST"); }
+
+    uint8_t getMethodByte() const override { return static_cast<uint8_t>(CompressionMethodByte::FSST); }
+
+    void updateHash(SipHash & hash) const override { getCodecDesc()->updateTreeHash(hash, /*ignore_aliases=*/true); }
+
+    static const int OUT_SIZE = 2281337;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp]: Magic constant, please add a comment behind it how it was calculated.

File: src/Compression/CompressionCodecFSST.cpp [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+
+#include <cstddef>
+#include <cstdint>
+#include <cstring>
+#include <stdexcept>
+#include <fsst.h>
+
+namespace DB
+{
+
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:23]: For my understanding: Why are we using the original FSST version and not the FSST12 variant?

File: src/Compression/CompressionCodecFSST.cpp [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+class CompressionCodecFSST : public ICompressionCodec
+{
+public:
+    explicit CompressionCodecFSST() { setCodecDescription("FSST"); }
+
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:28]: Could we mark the codec experimental for now? (`ICompressionCodec::isExperimental`)

File: contrib/fsst-cmake/CMakeLists.txt [hunk 5]
```diff
@@ -0,0 +1,19 @@
+option(ENABLE_FSST "Enable FSST (Fast Static Symbol Table)" ${ENABLE_LIBRARIES})
+
+if (NOT ENABLE_FSST)
```
Comment (rschu1ze) [inline contrib/fsst-cmake/CMakeLists.txt:3]: Note to myself: the AVX512 detection at the beginning of fsst_avx512.cpp looks x86-specific. Test what happens on ARM, possibly exclude fsst_avx512.cpp from compilation on non-x86 platforms.

File: src/Compression/CompressionCodecFSST.cpp [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+        splitDataByRows(reinterpret_cast<const unsigned char *>(source), str_in, len_in, source_size);
+        size_t rows_count{len_in.size()};
+
+        fsst_encoder_t * encoder = fsst_create(rows_count, len_in.data(), str_in.data(), 0);
+
+        size_t fsst_header_size = fsst_export(encoder, reinterpret_cast<unsigned char *>(dest));
+
+        size_t len_out[rows_count];
+        unsigned char * str_out[rows_count];
+        size_t header_size{fsst_header_size + sizeof(rows_count) + sizeof(len_out) + (sizeof(size_t) * len_in.size())};
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:50]: Minor: Let's use standard `=` initialization instead of use uniform initialization. The former is how it is done in the rest of the codebase. Please change all such places in this file.

File: src/Compression/CompressionCodecFSST.cpp [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+    uint8_t getMethodByte() const override { return static_cast<uint8_t>(CompressionMethodByte::FSST); }
+
+    void updateHash(SipHash & hash) const override { getCodecDesc()->updateTreeHash(hash, /*ignore_aliases=*/true); }
+
+    static constexpr int out_size = 2281337;
+
+protected:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override
+    {
+        std::vector<size_t> len_in;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:38]: Let's use human-readable names:
- `len_in` --> `string_ptrs`, `str_in` --> `string_lengths`
- In l. 48: `compressed_string_ptrs` and `compressed_string_lengths`.

(same in `decompress`, please)

File: src/Compression/CompressionCodecFSST.cpp [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+        std::vector<const unsigned char *> str_in;
+
+        splitDataByRows(reinterpret_cast<const unsigned char *>(source), str_in, len_in, source_size);
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:41]: If l. 39 would use `const char *`, we could remove all of the funny casting here and in the implementation of `splitDataByRows`.

File: src/Compression/CompressionCodecFSST.cpp [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        size_t rows_count{len_in.size()};
+
+        fsst_encoder_t * encoder = fsst_create(rows_count, len_in.data(), str_in.data(), 0);
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:44]: \+ `SCOPE_EXIT({ fsst_destroy(encoder); });`, otherwise we'll leak memory.

File: src/Compression/CompressionCodecFSST.cpp [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+
+        size_t fsst_header_size = fsst_export(encoder, reinterpret_cast<unsigned char *>(dest));
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:46]: `fsst_header_size` --> `symboltable_size` (otherwise this is too easy to confuse with `header_size`, l. 50). Let's apply the same change in `decompress`.

File: src/Compression/CompressionCodecFSST.cpp [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+
+        size_t len_out[rows_count];
+        unsigned char * str_out[rows_count];
+        size_t header_size{fsst_header_size + sizeof(rows_count) + sizeof(len_out) + (sizeof(size_t) * len_in.size())};
+        /* codec_header |(dest*) fsst_header(encoder) rows_count len_out len_in data */
+
+        if (fsst_compress(
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:53]: Let's use some human-readable formatting:

```cpp
const size_t num_compressed_strings = fsst_compress(encoder, string_count,
                                                    len_in.data(), str_in.data(),
                                                    out_size, reinterpret_cast<unsigned char *>(dest + header_size), len_out, str_out);
if (num_compressed_strings < string_count)
    throw std::runtime_error("FSST compression failed");
```

File: src/Compression/CompressionCodecFSST.cpp [hunk 12]
```diff
@@ -0,0 +1,144 @@
...
+                encoder,
+                rows_count,
+                len_in.data(),
+                str_in.data(),
+                out_size,
+                reinterpret_cast<unsigned char *>(dest + header_size),
+                len_out,
+                str_out)
+            < rows_count)
+            throw std::runtime_error("FSST compression failed");
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:63]: Throw a `LOGICAL_ERROR` instead, please (grep it in the codebase)

File: src/Compression/CompressionCodecFSST.cpp [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+        /* fsst_destroy(encoder); TODO(ebek): Понять почему вызывается деструктор какого-то левого кодека */
+
+        /* Copy prerequisites to dest */
+        memcpy(dest + fsst_header_size, &rows_count, sizeof(rows_count));
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:67]: For mental sanity, let's split this into three statements:

```cpp
dest += fsst_header_size;
memcpy(dest, &rows_count, sizeof(rows_count));
dest += sizeof(rows_count);
```

(likewise for the other memcpy-s)

File: src/Compression/CompressionCodecFSST.cpp [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+    static constexpr int out_size = 2281337;
+
+protected:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override
+    {
+        std::vector<size_t> len_in;
+        std::vector<const unsigned char *> str_in;
+
+        splitDataByRows(reinterpret_cast<const unsigned char *>(source), str_in, len_in, source_size);
+        size_t rows_count{len_in.size()};
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:42]: Codecs are so low-level in the stack that there is technically no notion of "rows" (and tables), there are only rows. Let's rename to "strings_count".

File: src/Compression/CompressionCodecFSST.cpp [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+                reinterpret_cast<const unsigned char *>(str),
+                out_size,
+                reinterpret_cast<unsigned char *>(dest));
+
+            str += lens[i];
+            dest += decompressed_size;
+        }
+    }
+
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override { return uncompressed_size + FSST_MAXHEADER; }
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:110]: Let's move l. 110-113 up to l. 31 so all "metadata" functions are in one place.

File: src/Compression/CompressionCodecFSST.cpp [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+
+private:
+    void splitDataByRows(
+        const unsigned char * data, std::vector<const unsigned char *> & rows, std::vector<size_t> & lens, UInt32 source_size) const
+    {
+        const unsigned char * end{data + source_size};
+
+        while (data != end)
+        {
+            UInt64 cur_len;
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:123]: I am pleasantly surprised that ClickHouse serializes String columns as is consecutive strings instead as two arrays (strings + length). That, of course, simplifies the job of `splitDataByRows`.

File: src/Compression/CompressionCodecFSST.cpp [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        /* Count data total compressed size without header */
+        UInt32 compressed_size{0};
+        for (size_t i = 0; i < rows_count; ++i)
+            compressed_size += len_out[i];
+        return static_cast<UInt32>(header_size) + compressed_size;
+    }
+
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override
+    {
+        UNUSED(uncompressed_size, source_size);
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:80]: Let's do instead in l. 78: `UInt32 /*uncompressed_size*/`)

File: src/Compression/tests/gtest_compressionCodec.cpp [hunk 18]
```diff
@@ -1303,4 +1304,41 @@ TEST(LZ4Test, DecompressMalformedInput)
     ASSERT_THROW(codec->decompress(source, source_size, memory.data()), Exception);
 }
 
+TEST(FSSTTest, CompressDecompress)
```
Comment (rschu1ze) [inline src/Compression/tests/gtest_compressionCodec.cpp:1307]: C++ unit tests are kind of underused in ClickHouse. Imho, all of what this test does can be achieved with a SQL-based test. The advantage of the latter is that it will run with all kinds of sanitizers in CI. 

Suggest to remove this test and extend the SQL test instead.

File: tests/queries/0_stateless/02973_fsst_code_test_data.sql [hunk 19]
```diff
@@ -0,0 +1,5 @@
+--Tags: no-fasttest
+
+CREATE TABLE table_fsst_codec (n String CODEC(FSST)) ENGINE = Memory;
+INSERT INTO table_fsst_codec VALUES ('Hello'), ('world'), ('!');
+SELECT * FROM table_fsst_codec;
```
Comment (rschu1ze) [inline tests/queries/0_stateless/02973_fsst_code_test_data.sql:5]: \+ `DROP TABLE`

**Review by rschu1ze — COMMENTED** [review]
I am curious how compression ration and performance compares to ZSTD?

You could load any of these datasets https://clickhouse.com/docs/en/getting-started/example-datasets and re-compress the string columns with different codecs, then do full-column scans over then so they are decompressed.

File: src/Compression/CompressionCodecFSST.cpp [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        memcpy(lens, source + fsst_header_size + sizeof(rows_count), sizeof(lens));
+        memcpy(len_in.data(), source + fsst_header_size + sizeof(rows_count) + sizeof(lens), sizeof(size_t) * len_in.size());
+        size_t header_size{fsst_header_size + sizeof(rows_count) + sizeof(lens) + sizeof(size_t) * len_in.size()};
+
+        const char * str{source + header_size};
+        for (size_t i = 0; i < rows_count; ++i)
+        {
+            dest = writeVarUInt(len_in[i], dest);
+
+            auto decompressed_size = fsst_decompress(
```
Comment (rschu1ze) [inline src/Compression/CompressionCodecFSST.cpp:98]: Docs in fsst.h say `If > size, the decoded output is truncated to size.`. Can this happen? Should we check?

**Comment by trololo23** [comment]
> I am curious how compression ration and performance compares to ZSTD?
> 
> You could load any of these datasets https://clickhouse.com/docs/en/getting-started/example-datasets and re-compress the string columns with different codecs, then do full-column scans over then so they are decompressed.

Hello! Here are the compression ratio on various datasets:

## [Trips](https://clickhouse.com/docs/en/getting-started/example-datasets/nyc-taxi) - pickup_ntaname and dropoff_ntaname columns:
- FSST 
trips_fsst
- STRING(LZ4)
trips_lz4
- LowCardinality
trips_low_cardinality
- ZSTD
trips_zstd

## [Amazon reviews](https://clickhouse.com/docs/en/getting-started/example-datasets/amazon-reviews) - review_body columns:
- FSST
amazon_fsst
- STRING(LZ4)
amazon_string
- ZSTD
amazon_zstd

## [Reddit comments](https://clickhouse.com/docs/en/getting-started/example-datasets/reddit-comments) - body column:
- FSST
redit_fsst
- STRING(LZ4)
redit_string
- ZSTD
redit_zstd

To summarize, fast is significantly inferior to zstd in terms of compression ratio, but in some cases (mainly on large rows) it surpasses LZ4. We will add the results of the performance measurement soon.

**Comment by rschu1ze** [comment]
Interesting, thanks. The comparison with LZ4 in terms of compression rate confirms the findings in [the FSST paper](https://raw.githubusercontent.com/cwida/fsst/master/fsstcompression.pdf). And of course, Zstd compresses really well...

I guess the true advantage of FSST as a _light-weight_ compression codec compared to the existing _heavy-weight_ codecs in ClickHouse is its ability for random access. Unfortunately, `ICompressionCodec::doDecompress` decompresses the entire block at once, so the existing API can't leverage random access.

@Pelanglene I am not sure how much more time you can/want to spend on this project, but there are two interesting directions to explore (suggested by @alexey-milovidov):

*Option 1*: Expand the `ICompressionCodec` interface for random access.

There are two sub-cases:
- _get_: decompress selected rows instead of the whole thing. This is useful for projection when most rows were removed by filters already. API-wise, one needs to pass in a bit vector into `doDecompress` (or an overload of `doDecompress`) to indicate the rows to decompress. FSST is well suited for that use case.
- _filter_: aka. this is filter pushdown to the codec level. Some codecs (including FSST) allow to evaluate predicates on compressed data, i.e. without decompression. As an example, the simplest case would be an equality filter on dictionary/domain-compresed (`LowCardinality` in ClickHouse) columns where one would translate x = 'abc' into the value id of x, and then scan the (compressed) value id vector for matches. The question is how to represent the filter ... I guess arbitrary operator + a list of arbitrary operand(s) will do the trick. This is as expressive as everything what the user can write in SQL. To take advantage of this, one also needs to create the filter in the query interpreter (more precisely, in query analysis) and carry it through all data reading interfaces down to the codec level.

*Option 2*: If a String (possibly `Nullable`, `Array(String)`) uses the FSST codec, then during deserialization, return a `ColumnFSST` instead of `ColumnString`. This is a new type of column that contains uncompressed bytes from one or more granules for subsequent lazy decompression. Only a few of its methods are implemented (filter, cut), whereas in almost all other cases, it must first be materialized into a full-fledged column.

**Comment by rschu1ze** [comment]
There are unfortunately too many unknown unknowns, e.g., what is the performance (compression/decompression) and size impact of ColumFSST, while the implementation is not finished. I still like the direction (FSSTs as internal representation for String/FixedString columns, as well as predicate pushdown to the codec level) but this needs more experimentation. I guess we should give it another try next year.

**Comment by alexey-milovidov** [comment]
@rschu1ze, the next year is today.
